ssl/os_port.h

```diff
@@ -102,60 +102,15 @@ void ax_wdt_feed();
 #define ax_array_read_u8(x, y) x[y]
 #else
 
-static inline uint8_t pgm_read_byte(const void* addr) {
-  register uint32_t res;
-  __asm__("extui    %0, %1, 0, 2\n"     /* Extract offset within word (in bytes) */
-      "sub      %1, %1, %0\n"       /* Subtract offset from addr, yielding an aligned address */
-      "l32i.n   %1, %1, 0x0\n"      /* Load word from aligned address */
-      "slli     %0, %0, 3\n"        /* Multiply offset by 8, yielding an offset in bits */
-      "ssr      %0\n"               /* Prepare to shift by offset (in bits) */
-      "srl      %0, %1\n"           /* Shift right; now the requested byte is the first one */
-      :"=r"(res), "=r"(addr)
-      :"1"(addr)
-      :);
-  return (uint8_t) res;     /* This masks the lower byte from the returned word */
-}
-
 #define ax_array_read_u8(x, y) pgm_read_byte((x)+(y))
 #endif //WITH_PGM_READ_HELPER
 
 #ifdef printf
 #undef printf
 #endif
 //#define printf(...)  ets_printf(__VA_ARGS__)
-#define PSTR(s) (__extension__({static const char __c[] PROGMEM = (s); &__c[0];}))
 #define PGM_VOID_P const void *
-static inline void* memcpy_P(void* dest, PGM_VOID_P src, size_t count) {
-    const uint8_t* read = (const uint8_t*)(src);
-    uint8_t* write = (uint8_t*)(dest);
-
-    while (count)
-    {
-        *write++ = pgm_read_byte(read++);
-        count--;
-    }
-
-    return dest;
-}
-static inline int strlen_P(const char *str) {
-    int cnt = 0;
-    while (pgm_read_byte(str++)) cnt++;
-    return cnt;
-}
-static inline int memcmp_P(const void *a1, const void *b1, size_t len) {
-    const uint8_t* a = (const uint8_t*)(a1);
-    uint8_t* b = (uint8_t*)(b1);
-    for (size_t i=0; i<len; i++) {
-        uint8_t d = pgm_read_byte(a) - pgm_read_byte(b);
-        if (d) return d;
-        a++;
-        b++;
-    }
-    return 0;
-}
-
```

Comment:
Likewise, although Arduino is build using a toolchain which includes these functions, it would be great if we can keep the code compatible with the "normal" toolchain as well. Can be done by means of an extra macro set in the makefile.

Reply:
Done, by doing a test build in the Makefile to see if the _P functions are already present, and if so set a flag to #ifdef them out.